### PR TITLE
#4586: Clear Results feature in Notebooks

### PR DESCRIPTION
**Pull request by rajmusuku**
This feature will allow us to clear outputs of all the code cells exist in a Notebook. Implementation is driven by an action. Right now,  'Clear Results' button is placed in tool bar. Let me know the feedback if something else need to be added/modified to this feature.

### CODE REVIEW

**Comment by rajmusuku** [comment]
Added integration tests by including extensible method for clear outputs. Please review it and let me know.

File: extensions/integration-tests/src/notebook.test.ts [hunk 1]
```diff
@@ -36,6 +36,26 @@ if (context.RunTest) {
 			assert(result === '2', `Expected: 2, Acutal: '${result}'`);
 		});
 
+		test('Clear all outputs - Python3 notebook ', async function () {
+			let notebook = await openNotebook(pySparkNotebookContent, pythonKernelMetadata);
+			//Check if at least one cell with output
+			let cellWithOutputs = notebook.document.cells.find(cell => cell.contents && cell.contents.outputs && cell.contents.outputs.length > 0);
```
Comment (chlafreniere) [inline extensions/integration-tests/src/notebook.test.ts]: there should be an assert here to ensure the cellWithOutputs isn't undefined. also please add some more logging (take a look at what @YurongHe did earlier today)

File: src/sql/parts/notebook/notebook.component.ts [hunk 2]
```diff
@@ -371,6 +372,8 @@ export class NotebookComponent extends AngularDisposable implements OnInit, OnDe
 		let addTextCellButton = new AddCellAction('notebook.AddTextCell', localize('text', 'Text'), 'notebook-button icon-add');
 		addTextCellButton.cellType = CellTypes.Markdown;
 
+		let clearResultsButton = new ClearAllOutputsAction('notebook.ClearAllOutputs', localize('clearResults', 'Clear Results'), 'notebook-button icon-clear-results');
```
Comment (chlafreniere) [inline src/sql/parts/notebook/notebook.component.ts]: we should be consistent with naming. Please consider renaming ClearAllOutputsAction to ClearAllResultsAction

File: src/sql/parts/notebook/notebookActions.ts [hunk 3]
```diff
@@ -53,6 +54,19 @@ export class AddCellAction extends Action {
 	}
 }
 
+
+// Action to clear outputs of all code cells.
```
Comment (chlafreniere) [inline src/sql/parts/notebook/notebookActions.ts]: outputs --> results

File: extensions/integration-tests/src/notebook.test.ts (hunk 1 above)
Comment (rajmusuku) [reply extensions/integration-tests/src/notebook.test.ts]: Will add. Will check with Yurong, however, I wonder what more logging is needed here as it's 4 lines of code and we have two logging statements, one test and the other one is assert.

File: src/sql/parts/notebook/notebook.component.ts (hunk 2 above)
Comment (rajmusuku) [reply src/sql/parts/notebook/notebook.component.ts]: Agree from a consistency point of view. So far, we have been follwing the convention of defining code cell's result as 'Output' across our code base. Looking at output classess, interfaces and variables, I decided to maintain the same naming convension here as well. Howver, the label can be any thing, in this case it happened to be "Clear Results", but from a coding perspective I am more inclined to keep it as 'Output' inline with other output objects. I am happy to change it if it's really needed.

File: src/sql/parts/notebook/notebook.component.ts [hunk 4]
```diff
@@ -479,4 +483,19 @@ export class NotebookComponent extends AngularDisposable implements OnInit, OnDe
 		}
 	}
 
+	public async clearAllOutputs(): Promise<boolean> {
+		try {
+			await this.modelReady;
+			this._model.cells.forEach(cell => {
+				if (cell.cellType === CellTypes.Code) {
+					(cell as CellModel).clearOutputs();
```
Comment (YurongHe) [inline src/sql/parts/notebook/notebook.component.ts]: what happen if cell is running state?

File: src/sql/parts/notebook/notebookActions.ts (hunk 3 above)
Comment (rajmusuku) [reply src/sql/parts/notebook/notebookActions.ts]: Same as above comment as we are using 'output' in codebase

File: extensions/integration-tests/src/notebook.test.ts [hunk 5]
```diff
@@ -43,6 +43,36 @@ if (context.RunTest) {
...
+			let notebook = await openNotebook(sqlNotebookContent, sqlKernelMetadata);
+			let cellWithOutputs = notebook.document.cells.find(cell => cell.contents && cell.contents.outputs && cell.contents.outputs.length > 0);
+			console.log("Before clearing cell outputs");
+			if (cellWithOutputs) {
+				let clearedOutputs = await notebook.clearAllOutputs();
+				assert(clearedOutputs, 'Outputs of all the code cells from SQL notebook should be cleared');
+				console.log("After clearing cell outputs");
+			}
+			else{
+				assert('Could not find notebook cells with outputs');
```
Comment (Charles-Gagnon) [inline extensions/integration-tests/src/notebook.test.ts]: This doesn't seem correct - it's asserting the string (which is always going to pass)

File: extensions/integration-tests/src/notebook.test.ts [hunk 6]
```diff
@@ -36,6 +36,26 @@ if (context.RunTest) {
 			assert(result === '2', `Expected: 2, Acutal: '${result}'`);
 		});
 
+		test('Clear all outputs - Python3 notebook ', async function () {
+			let notebook = await openNotebook(pySparkNotebookContent, pythonKernelMetadata);
+			//Check if at least one cell with output
+			let cellWithOutputs = notebook.document.cells.find(cell => cell.contents && cell.contents.outputs && cell.contents.outputs.length > 0);
+			if (cellWithOutputs) {
```
Comment (YurongHe) [inline extensions/integration-tests/src/notebook.test.ts]: Just wondering if we can check celloutput is empty.

**Comment by michelmetran** [comment]
When using ipynb files with Python, it is possible to add the following content to the `.gitattributes` file:
```
# Filters
*.ipynb filter=nbstripout
*.ipynb diff=ipynb
```

This way, when pushing to a git repository, all notebook output will be automatically cleaned up.

Is something similar possible in Azure Data Studio?

**Pull request merged**